Code/GraphMol/Chirality.cpp

```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
   return res;
 }
 
-// returns true if the atom is allowed to have stereochemistry specified
-bool checkChiralAtomSpecialCases(ROMol &mol, const Atom *atom) {
-  PRECONDITION(atom, "bad atom");
-
+// finds all possible chiral special cases.
+// at the moment this is just candidates for ring stereochemistry
+void findChiralAtomSpecialCases(ROMol &mol,
+                                boost::dynamic_bitset<> &possibleSpecialCases) {
+  PRECONDITION(possibleSpecialCases.size() >= mol.getNumAtoms(),
+               "bit vector too small");
+  possibleSpecialCases.reset();
   if (!mol.getRingInfo()->isInitialized()) {
     VECT_INT_VECT sssrs;
     MolOps::symmetrizeSSSR(mol, sssrs);
   }
-
-  const RingInfo *ringInfo = mol.getRingInfo();
-  if (ringInfo->numAtomRings(atom->getIdx()) &&
-      atomIsCandidateForRingStereochem(mol, atom)) {
-    // the atom is in a ring, so the "chirality" specification may actually
-    // be handling ring stereochemistry.
-
-    // check for another chiral tagged
-    // atom without stereochem in this atom's rings:
+  boost::dynamic_bitset<> atomsSeen(mol.getNumAtoms());
+  boost::dynamic_bitset<> atomsUsed(mol.getNumAtoms());
+  boost::dynamic_bitset<> bondsSeen(mol.getNumBonds());
+
+  for (ROMol::AtomIterator ait = mol.beginAtoms(); ait != mol.endAtoms();
+       ++ait) {
+    const Atom *atom = *ait;
+    if (atomsSeen[atom->getIdx()]) continue;
+    if (atom->getChiralTag() == Atom::CHI_UNSPECIFIED ||
+        atom->hasProp(common_properties::_CIPCode) ||
+        !mol.getRingInfo()->numAtomRings(atom->getIdx()) ||
+        !atomIsCandidateForRingStereochem(mol, atom)) {
+      continue;
+    }
+    // do a BFS from this ring atom along ring bonds and find other
+    // stereochemistry candidates.
+    std::list<const Atom *> nextAtoms;
```

Comment:
can we please do the optimization stuff later? I'm concerned that this is big and complex enough that we'll both end up spending weeks on the review if we convolve correctness and performance (which is no worse than what was there before)